### PR TITLE
Replace Reflection Calls with MethodHandles

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR prototypes replacing reflective invocations with `MethodHandle` invocations. `MethodHandle`s provide a better performance experience where their permissions requirements are checked during constructions instead of on a per call basis with reflection. When using a `SecurityManager` this reduces the amount of time spent validating the call and `MethodHandle`s have better support for inlining and other performance benefits the JVM can apply.

### CODE REVIEW

**Review by joshfree — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ResponseConstructorsCache.java [hunk 1]
```diff
@@ -10,58 +10,103 @@
...
+    private MethodHandle locateResponseConstructor(Class<?> responseClass) {
+        /*
+         * If we were able to write this using Java 9+ code.
+         *
+         * First check if the response class's module is exported to all unnamed modules. If it is we will use
+         * MethodHandles.publicLookup() which is meant for creating MethodHandle instances for publicly accessible
+         * classes.
+         */
+        MethodHandles.Lookup lookupToUse;
+        if (responseClass.getModule().isExported("")) {
```
Comment (hemanttanwar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ResponseConstructorsCache.java:82]: `responseClass` : Does this needs to be checked for null ?

**Review by hemanttanwar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ResponseConstructorsCache.java [hunk 2]
```diff
@@ -71,40 +167,77 @@
...
+        }
+    }
+
+    private static void addModuleRead(Object module, ClientLogger logger) {
+        try {
+            ADD_MODULE_READ.invoke(CORE_MODULE, module);
+        } catch (Throwable throwable) {
+            throw logger.logExceptionAsError(Exceptions.propagate(throwable));
+        }
+    }
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ResponseConstructorsCache.java:200]: For these methods, catch/log and throw may not be required. The static block that calls these methods, ignores these exceptions and just logs a verbose message with the exception. So, logging them here as an error can be misleading. Calls from `locateResponseConstructor` method should handle the exception and fallback to default lookup.

**Comment by JonathanGiles** [comment]
@alzimmermsft Let's get this finished off and merged. It's a shame to leave it hanging around!

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/ReflectionUtils.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     * If Java 9 or above is being used this will return a {@link MethodHandles.Lookup} based on whether the module
+     * containing the {@code targetClass} exports the package containing the class. Otherwise, the {@link
+     * MethodHandles.Lookup} associated to {@code com.azure.core} will attempt to read the module containing {@code
+     * targetClass}.
+     *
+     * @param targetClass The {@link Class} that will need to be reflectively accessed.
+     * @return The {@link MethodHandles.Lookup} that will allow {@code com.azure.core} to access the {@code targetClass}
+     * reflectively.
+     */
+    public static MethodHandles.Lookup getLookupToUse(Class<?> targetClass) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/ReflectionUtils.java:79]: Check this with `Throwable` and force callers to handle the exception appropriately, either swallow and use a fallback path or propagate.

**Review by srnagar — APPROVED** [review]
LGTM!

**Pull request merged**